docs/spec/overload.rst

```diff
@@ -99,8 +109,376 @@ variable is not sufficient.
 
 Another important difference between type variables such as ``AnyStr``
 and using ``@overload`` is that the prior can also be used to define
-constraints for generic class type parameters.  For example, the type
+constraints for generic class type parameters. For example, the type
 parameter of the generic class ``typing.IO`` is constrained (only
 ``IO[str]``, ``IO[bytes]`` and ``IO[Any]`` are valid)::
 
   class IO(Generic[AnyStr]): ...
+
+
+Invalid overload definitions
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Type checkers should enforce the following rules for overload definitions.
+
+At least two ``@overload``-decorated definitions must be present. If only
+one is present, an error should be reported.
+
+An overload implementation should be present for all overload function
+definitions. Type checkers should report an error or warning if an
+implementation is missing. Overload definitions within stub files,
+protocols, and abstract base classes are exempt from this check.
+
+If an overload is decorated with ``@staticmethod`` or ``@classmethod``,
+all overloads must be similarly decorated. The implementation,
+if present, must be decorated in the same manner. Similarly, if one overload
+is ``async``, all overloads must be ``async`` as well as the implementation.
+Type checkers should report an error if these conditions are not met.
+
+If one or more overloads are decorated with ``@final`` but the
+implementation is not, an error should be reported.
+
+
+Implementation consistency
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+If an overload implementation is defined, type checkers should validate
+that its signature is consistent with all of its associated overload
+signatures. The implementation should accept all potential argument lists
+that are accepted by the overloads and should produce all potential return
+types produced by the overloads. In typing terms, this means the input
+signature of the implementation should be assignable to the input signatures
+of all overloads, and the return type of all overloads should be assignable to
+the return type of the implementation.
+
+If the implementation is inconsistent with its overloads, a type checker
+should report an error::
+
+  @overload
+  def func(x: str, /) -> str: ...
+  @overload
+  def func(x: int) -> int: ...
+
+  # This implementation is inconsistent with the second overload
+  # because it does not accept a keyword argument ``x`` and the
+  # the overload's return type ``int`` is not assignable to the
+  implementation's return type ``str``.
+  def func(x: int | str, /) -> str:
+    return str(x)
+
+Overlapping overloads
+^^^^^^^^^^^^^^^^^^^^^
+
+If two overloads can accept the same set of arguments but have
+different return types, the overloads are said to "partially overlap".
```

Comment:
"different" isn't quite the right condition, since we allow:
```
from typing import *
@overload
def func(x: Literal[0]) -> Literal[0]: ...
@overload
def func(x: int) -> int: ...
```

Reply:
Yes, you're correct. I'll correct the language.